File: go/node.go
package immgo

import (
	"fmt"
)

// Node represents a node in a tree with some arbitrary data of type `T`
// attached.
type Node[T any] struct {
	data T

	firstChild *Node[T]
	nextSibling *Node[T]
}

func NewNode[T any]() *Node[T] {
	// defaults are fine
	return &Node[T] {}
}

// Traverse this node and all children, calling `f` on each.
func (this *Node[T]) Traverse(f func(*Node[T])) {
	f(this)

	children := this.Children()

	for _, child := range children {
		child.Traverse(f)
	}
}

func (this *Node[T]) Children() []*Node[T] {
	r := []*Node[T]{}

	curr := this.firstChild
	for {
		if curr == nil {
			break
		}

		r = append(r, curr)
		curr = curr.nextSibling
	}

	return r
}

func (this *Node[T]) RemoveChildAt(index int) error {
	children := this.Children()

	if index < 0 || index >= len(children) {
		return fmt.Errorf("Index past bounds of children")
	}

	if index == 0 {
		this.firstChild = this.firstChild.nextSibling
	} else {
		children[index-1].nextSibling = children[index].nextSibling
	}

	return nil
}

func (this *Node[T]) InsertChildAt(child *Node[T], index int) error {
	children := this.Children()

	if index < 0 || index > len(children) {
		return fmt.Errorf("Index past bounds of children")
	}

	if index == 0 {
		child.nextSibling = this.firstChild
		this.firstChild = child
	} else {
		child.nextSibling = children[index-1].nextSibling
		children[index-1].nextSibling = child
	}

	return nil
}


func (this *Node[T]) AppendChild(child *Node[T]) {
	lastChild := this.LastChild()
	if lastChild == nil {
		this.firstChild = child
	} else {
		lastChild.nextSibling = child
	}
}

func (this *Node[T]) HasChild(node *Node[T]) bool {
	children := this.Children()
	for _, child := range children {
		if child == node {
			return true
		}
	}

	return false
}

func (this *Node[T]) LastChild() *Node[T] {
	if this.firstChild == nil {
		return nil
	}

	curr := this.firstChild
	for {
		if curr.nextSibling == nil {
			return curr
		}

		curr = curr.nextSibling
	}
}

type Attributes = map[string]interface{}
type EventHandlers = map[string]func(event interface{})

// Properties include attributes, which are keyed by attribute name and can be
// any time (though are usually strings), and event handlers, which are keyed by
// event type.
type Properties struct {
	Attributes map[string]interface{}
	EventHandlers map[string]func(event interface{})
}

type ElementDescription struct {
	Key string
	Kind string
	Properties Properties

	// set by the framework
	// todo: move to rendernodedata
	fullkey string
}

type RenderNodeData struct {
	description ElementDescription
	match *ShadowNode
}

type RenderNode = Node[RenderNodeData]

func NewRenderNode(key, kind string, properties Properties) *RenderNode {
	node := NewNode[RenderNodeData]()
	node.data = RenderNodeData {
		description: ElementDescription {
			Key: key,
			Kind: kind,
			Properties: properties,
		},
		match: nil,
	}

	return node
}

type ShadowNode = Node[ElementDescription]

func NewShadowNode(key, kind string, properties Properties) *ShadowNode {
	node := NewNode[ElementDescription]()
	node.data = ElementDescription {
		Key: key,
		Kind: kind,
		Properties: properties,
	}

	return node
}

func NewShadowNodeForRenderNode(renderNode *RenderNode) *ShadowNode {
	r := NewShadowNode(
		renderNode.data.description.Key,
		renderNode.data.description.Kind,
		renderNode.data.description.Properties,
	)
	r.data.fullkey = renderNode.data.description.fullkey
	return r
}

File: py/immpy/node.py
from __future__ import annotations

from typing import TypeVar, Generic, Optional, List, Dict, Any
from dataclasses import dataclass

from mypy_utils import unwrap

T = TypeVar('T')

class Node(Generic[T]):
    def __init__(self, value: T):
        self.data = value
        self.first_child: Optional[Node[T]] = None
        self.next_sibling: Optional[Node[T]] = None

    def children(self) -> List[Node[T]]:
        r = []
        curr = self.first_child
        while curr is not None:
            r.append(curr)
            curr = curr.next_sibling

        return r

    def remove_child_at(self, index: int):
        children = self.children()
        if index < 0 or index >= len(children):
            raise IndexError

        if index == 0:
            self.first_child = unwrap(self.first_child).next_sibling
        else:
            children[index-1].next_sibling = children[index].next_sibling

    def insert_child_at(self, child: Node[T], index: int):
        children = self.children()

        if index < 0 or index > len(children):
            raise IndexError

        if index == 0:
            child.next_sibling = self.first_child
            self.first_child = child
        else:
            child.next_sibling = children[index-1].next_sibling
            children[index-1].next_sibling = child

    def append_child(self, child: Node[T]):
        last_child = self.last_child()
        if last_child is None:
            self.first_child = child
        else:
            last_child.next_sibling = child

    def has_child(self, node: Node[T]) -> bool:
        children = self.children()
        return any([child == node for child in children])

    def last_child(self) -> Optional[Node[T]]:
        if self.first_child is None:
            return None

        curr = self.first_child
        while curr is not None:
            if curr.next_sibling is None:
                return curr

            curr = curr.next_sibling


Properties = Dict[str, Any]

@dataclass
class RenderNodeData:
    key: str
    kind: str
    properties: Properties

RenderNode = Node[RenderNodeData]

ShadowNodeData = RenderNodeData
ShadowNode = Node[ShadowNodeData]

def create_shadow_node_for_render_node(render_node: RenderNode):
    return ShadowNode(
            ShadowNodeData(
                key=render_node.data.key,
                kind=render_node.data.kind,
                properties=render_node.data.properties,
            )
    )
